text/0011-gem-checksum-verification.md

```diff
@@ -0,0 +1,184 @@
+
+Example:
+
+```
+$ bundle pristine
+Installing rake 13.0.6
+…
+Installing rspec 3.12.0
+42 gems are missing checksums. Source integrity could not be verified.
+Use `bundle lock` to add checksums to Gemfile.lock.
+$ bundle lock --update
+Writing lockfile to path/to/Gemfile.lock
+$ bundle pristine
+Installing rake 13.0.6 (verified)
+…
+Installing rspec 3.12.0 (verified)
+$
+```
+
+If a gem that is about to be installed does not match the checksum, an error is generated.
+
+Example:
+
+```
+$ bundle install
+Installing rake 13.0.6
+Bundler cannot continue installing rake-13.0.6.
+The checksum for the downloaded `rake-13.0.6.gem` does not match the known checksum for the gem. 
+
+Expected: sha256-814828c34f1315d7e7b7e8295184577cc4e969bad6156ac069d02d63f58d82e8 from: Gemfile.lock:21:1 CHECKSUMS rake (13.0.6)
+Downloaded: sha256-69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b69b6 from:
+path/to/rake-13.0.6.gem SHA256 hexdigest
+
+To resolve this issue:
+Delete the downloaded gem files referenced above and run `bundle install`
+
+If you are sure that the downloaded gem is correct, you can run…
+(RFC TODO: Currently undecided on what the command should be here to allow unverified gems.)
+`bundle install --skip-verification` OR
+`bundle config set unverified ignore` OR
+Remove the offending line from CHECKSUMS in Gemfile.lock
+```
+
+Certain checksums will always be unavailable because the source does not provide a checksum. When a checksum is not available, a line can be added to the Gemfile.lock CHECKSUMS section that explicitly indicates that no checksum is recorded for that gem version. Future installs of gems with intentionally ignored checksums should not error or warn when the gem is installed. The **(RFC TODO)** command will allow idicating that a gem version should ignore checksum verifications. Alternatively, the SHA256 digest of the gem at install time can be recorded allowing the user to verify future installs do not deviate from the recorded checksum.
+
+Platforms may pose a problem for Gemfile.lock checksums. If a user has not added one of the platforms where they bundle their app, then checksum verification could be skipped. In order to avoid this security vulnerability, the following should happen automatically.
+
+1. When a gem is installed and no checksum is available in the lockfile, a warning should be printed indicating that unverified gems were installed.
+2. When a gem is installed where a checksum is available for the same gem on a different platform, an error should be raised. (i.e. if nokogiri for darwin has a checksum, but nokogiri for musl does not, then an error should be raised).
```

Comment:
this will be a breaking change to how bundler handles formats, and I would anticipate user backlash against this. (doubly so if, as above, it happens even when a checksum comes from the remote). Currently, you can transparent add a platform to the lock file via `bundle install`, and it is only disallowed under `--deployment` and `--frozen`

Reply:
I agree. I'll remove this line. I think I was imagining a frozen gemfile when I wrote this, but then I wrote the next like 3. which deals with frozen gemfiles.